server/server.go

```diff
@@ -62,3 +67,94 @@ func (s *Server) RenderingRegistrationDetails(w http.ResponseWriter, r *http.Req
 		return
 	}
 }
+
+func (s *Server) HomePageHandler(w http.ResponseWriter, r *http.Request) {
+	sessionID := GetSessionIDFromCookie(r)
+	session, err := s.app.SessionService.ValidateSession(sessionID)
+
+	var username string
+
+	if err == nil {
+		user, err := s.app.UserService.GetUserByID(session.UserId)
+		if err == nil {
+			username = user.Username
+		}
+	}
+	tmpl, _ := template.ParseFiles("views/homepage.html")
+	err = tmpl.Execute(w, struct {
+		Username string
+	}{username})
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (s *Server) LoginFormHandler(w http.ResponseWriter, r *http.Request) {
+	sessionID := GetSessionIDFromCookie(r)
+	_, err := s.app.SessionService.ValidateSession(sessionID)
+
+	if err != nil {
+		tmpl, _ := template.ParseFiles("views/login.html")
+		err := tmpl.Execute(w, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		return
+	}
+
+	http.Redirect(w, r, "/", http.StatusFound)
+}
+
+func (s *Server) LoginHandler(w http.ResponseWriter, r *http.Request) {
+	email := r.FormValue("email")
+	password := r.FormValue("password")
+
+	user, err := s.app.UserService.ValidateUserCredentials(email, password)
+	if err != nil {
+		tmpl, err := template.ParseFiles("views/login.html")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		data := struct {
+			ErrorMessage string
+		}{
+			ErrorMessage: "Invalid email or password",
+		}
+		err = tmpl.Execute(w, data)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		return
+	}
+
+	sessionID, err := s.app.SessionService.CreateSession(user.UserId, time.Now().Add(app.SessionDurationInHours*time.Hour))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	SetSessionCookie(w, sessionID)
+	http.Redirect(w, r, "/", http.StatusFound)
+}
+
+func SetSessionCookie(w http.ResponseWriter, sessionID string) {
+	http.SetCookie(w, &http.Cookie{
```

Comment:
The cookie is not still secure. Please make the cookie `Same-Site` strict, http-only and secure.